src/Conversion/ONNXToKrnl/Tensor/Unique.cpp

```diff
@@ -0,0 +1,194 @@
+  ///
+  /// sorted unique subtensors:
+  /// B: [[0, 1], [0, 1]],
+  ///    [[1, 1], [1, 1]],
+  ///    [[2, 1], [2, 1]].
+  ///
+  /// output_Y is constructed from B:
+  /// [[[0. 1.], [1. 1.], [2. 1.]],
+  /// [[0. 1.], [1. 1.], [2. 1.]]]
+  ///
+  /// output_indices is to map from B to A:
+  /// [1, 0, 2]
+  ///
+  /// output_inverse_indices is to map from A to B:
+  /// [1, 0, 2, 0]
+  ///
+  /// output_counts = [2 1 1]
+  ///
+
+  LogicalResult matchAndRewrite(Operation *op, ArrayRef<Value> operands,
+      ConversionPatternRewriter &rewriter) const final {
+    ONNXTopKOpAdaptor operandAdaptor(operands, op->getAttrDictionary());
+    ONNXUniqueOp uniqueOp = llvm::cast<ONNXUniqueOp>(op);
+    Location loc = op->getLoc();
+    MultiDialectBuilder<KrnlBuilder, IndexExprBuilderForKrnl, MathBuilder,
+        MemRefBuilder>
+        create(rewriter, loc);
+    IndexExprScope scope(create.krnl);
+    ONNXUniqueOpShapeHelper shapeHelper(op, operands, &create.krnlIE);
+    Value X = operandAdaptor.getX();
+    ArrayRef<int64_t> xShape = getShape(X.getType());
+    Type elementType = X.getType().cast<MemRefType>().getElementType();
+    int64_t rank = create.krnlIE.getShapedTypeRank(X);
+    int64_t sorted = operandAdaptor.getSorted();
+    std::optional<int64_t> optionalAxis = uniqueOp.getAxis();
+    //
+    // Get axis value as a positive integer if axis attribute is specified.
+    // If set "-1" if the axis attribute is not specified.
+    //
+    int64_t axis = -1;
+    if (optionalAxis.has_value()) {
+      axis = optionalAxis.value();
+      axis = axis < 0 ? axis + rank : axis;
+      assert(axis >= 0 && axis < rank && "axis is out of bound");
+    }
+    //
+    // Emit a Unique call to get the outputs to count unique
+    // slices(subtensors) of X along axis at first.
+    //
+    Type indexTy = rewriter.getIndexType();
+    Value iZero = create.math.constantIndex(0);
+    Value uniqueCount = create.mem.alloca(MemRefType::get({}, indexTy));
+    create.krnl.store(iZero, uniqueCount, {});
+    Value noneValue;
+    emitArgUnique(rewriter, loc, uniqueCount, X, axis, /*sorted=*/sorted,
+        noneValue, noneValue, noneValue, noneValue, /*count_only=*/true);
+    //
+    // Calculate shapes of output Tensors
+    //
+    Value total = create.krnl.load(uniqueCount, {});
+    NonAffineIndexExpr totalDimExpr = DimIndexExpr(total);
+    DimsExpr outputYDims;
+    DimsExpr outputIndexDims;
+    DimsExpr outputInverseIndexDims;
+    if (axis < 0) {
+      outputYDims.emplace_back(totalDimExpr);
+      outputIndexDims.emplace_back(totalDimExpr);
+      DimIndexExpr inputDimExpr = LiteralIndexExpr(xShape[0]);
+      for (int64_t i = 1; i < rank; i++) {
+        inputDimExpr = inputDimExpr * LiteralIndexExpr(xShape[i]);
+      }
+      outputInverseIndexDims.emplace_back(inputDimExpr);
+    } else {
+      for (int64_t i = 0; i < rank; i++) {
+        DimIndexExpr tDimExpr = LiteralIndexExpr(xShape[i]);
+        if (i == axis)
```

Comment:
To me, most the work should be done in computeShape. Only thing needed to be fixed here is to set the tDimExpr to the right dimension. You didn't do that way because there is no way to represent the unknown value (uniqueCount) with IndexExpr. Is that the reason?

Reply:
> You didn't do that way because there is no way to represent the unknown value (uniqueCount) with IndexExpr. Is that the reason?

Probably yes according to my understanding.
This part generates code to calculate shapes of output buffers to be used for their buffer allocation before calling the second krnl.call according to the actual number of unique slices (uniqueCount), which is counted by the first krnl.call.  I am not very sure if the current way is good, and your comments are very welcome.

```
    //
    // Calculate shapes of output Tensors
    //
    Value total = create.krnl.load(uniqueCount, {});
    NonAffineIndexExpr totalDimExpr = DimIndexExpr(total);
    DimsExpr outputYDims;
    DimsExpr outputIndexDims;
    DimsExpr outputInverseIndexDims;
    if (axis < 0) {
      outputYDims.emplace_back(totalDimExpr);
      outputIndexDims.emplace_back(totalDimExpr);
      DimIndexExpr inputDimExpr = LiteralIndexExpr(xShape[0]);
      for (int64_t i = 1; i < rank; i++) {
        inputDimExpr = inputDimExpr * LiteralIndexExpr(xShape[i]);
      }
      outputInverseIndexDims.emplace_back(inputDimExpr);
    } else {
      for (int64_t i = 0; i < rank; i++) {
        DimIndexExpr tDimExpr = LiteralIndexExpr(xShape[i]);
        if (i == axis)
          tDimExpr = totalDimExpr;
        outputYDims.emplace_back(tDimExpr);
      }
      outputIndexDims.emplace_back(totalDimExpr);
      outputInverseIndexDims.emplace_back(LiteralIndexExpr(xShape[axis]));
    }
```